src/Agent.Plugins/PipelineCache/TarUtils.cs

```diff
@@ -0,0 +1,202 @@
+                        // readStdOut and readStdError should only fail if the process dies
+                        // processTcs.Task cannot fail as we only call SetResult on processTcs
+                        await Task.WhenAll(readStdOut, readStdError, processTcs.Task);
+                    }
+
+                    int exitCode = await processTcs.Task;
+
+                    if (exitCode == 0)
+                    {
+                        context.Output($"Process exit code: {exitCode}");
+                        foreach (string line in output)
+                        {
+                            context.Output(line);
+                        }
+                    }
+                    else
+                    {
+                        throw new Exception($"Process returned non-zero exit code: {exitCode}");
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Delete archive file.
+                    if (File.Exists(archiveFile))
+                    {
+                        File.Delete(archiveFile);
+                    }
+                    foreach (string line in output)
+                    {
+                        context.Error(line);
+                    }
+                    ExceptionDispatchInfo.Capture(e).Throw();
+                }
+            }
+            return archiveFile;
+        }
+
+        public async Task<bool> DownloadTar(AgentTaskPluginExecutionContext context, DedupManifestArtifactClient dedupManifestClient, DedupIdentifier dedupId, string targetDirectory, CancellationToken cancellationToken)
+        {
+            var processTcs = new TaskCompletionSource<int>();
+
+            using (var cancelSource = new CancellationTokenSource())
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cancelSource.Token))
+            using (var process = new Process())
+            {
+                process.StartInfo.FileName = isWindows ? "7z" : "tar";
+                process.StartInfo.Arguments = isWindows ? $"x -si -aoa -o{targetDirectory} -ttar" : $"-xf - -C {targetDirectory}";
+                process.StartInfo.UseShellExecute = false;
+                process.StartInfo.RedirectStandardInput = true;
+                process.StartInfo.RedirectStandardOutput = true;
+                process.StartInfo.RedirectStandardError = true;
+                process.EnableRaisingEvents = true;
+                process.Exited += (sender, args) =>
+                {
+                    cancelSource.Cancel();
+                    processTcs.SetResult(process.ExitCode);
+                };
+
+                context.Info($"Starting '{process.StartInfo.FileName}' with arguments '{process.StartInfo.Arguments}'...");
+                process.Start();
+
+                var output = new List<string>();
+
+                Task readLines(string prefix, StreamReader reader) => Task.Run(async () =>
```

Comment:
see comment above (in upload) ... re: is the value of these extra stdout/err tasks worth it? (i.e. worth the added code complexity and potential for problems)